python-checks/src/main/java/org/sonar/python/checks/DictionaryDuplicateKeyCheck.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.tree.DictionaryLiteral;
+import org.sonar.plugins.python.api.tree.DictionaryLiteralElement;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.KeyValuePair;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.NumericLiteral;
+import org.sonar.plugins.python.api.tree.StringElement;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S5780")
+public class DictionaryDuplicateKeyCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.DICTIONARY_LITERAL, ctx -> {
+      DictionaryLiteral dictionaryLiteral = (DictionaryLiteral) ctx.syntaxNode();
+      Set<Integer> issueIndexes = new HashSet<>();
+      for (int i = 0; i < dictionaryLiteral.elements().size(); i++) {
```

Comment:
The handler in `initialize` in combination with the `for`-loop in
  `findIdenticalKeys` gives quadratic runtime in the number of elements of
  the dictionary.
  That seems a bit surprising for an "obviously O(n)" task.
  If someone decides
  to use a dictionary with few thousand elements, it might already hurt
  the performance, and in larger examples, it might freeze the analyzer altogether.

  Ideally, one might split up the check into two parts:
  First - normalization, i.e. bringing each key into some canonical form,
  with a well-defined hash code and easy comparison. Second - actual
  comparison for equivalence: two keys are equivalent w.r.t Python's `==`
  iff their normalized versions are equal in Java. That would give a
  linear algorithm, and the size of the analyzed dictionary wouldn't matter.

  I see that this whole rule relies on `CheckUtils.areEquivalent`, which does not give any normalized, easily hashable / comparable values, so that might be difficult to fix quickly. But maybe one might add a condition that we, for now, don't inspect dictionaries with more than, say, 100 entries, just to make sure that the analyzer doesn't freeze suddenly once confronted with longer dictionaries.

Reply:
That's a good point.
As you said I'm not sure it's a trivial task to normalize each key, especially from the current implementation.
I'm also not sure about the performance impacts in real-life scenarios, as I don't know how common big dictionary literals are. In doubt, a quick fix would be to disable the rule for big dictionaries indeed, but I guess it would only make sense if we actually face performance issues due to this. What's your opinion on this @andrea-guarino-sonarsource ?

Reply:
I believe that's a fair point. I'm fine with excluding large dictionaries as a first baby step.
If it makes sense and we believe there is value analyzing large dictionaries, we can improve the rule later.

Reply:
Indeed, I also looked a bit on the idea to normalize the keys and it doesn't seem straightforward due to the use of `BigDecimal` (as hashing those directly doesn't work due to the scale being part of the `equals` check), so I excluded dictionary literals of more than 100 elements in af74d2bb08b0c29d5d5b4ec9ca09a5b399d3dc90.